vault/data_source_auth_backends.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+)
+
+func authBackendsDataSource() *schema.Resource {
+	return &schema.Resource{
+		Read: ReadWrapper(authBackendsDataSourceRead),
+		Schema: map[string]*schema.Schema{
+			"paths": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "The auth backend mount points.",
+			},
+			"type": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The name of the auth backend.",
+			},
+			"accessors": {
+				Type:        schema.TypeList,
+				Computed:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "The accessors of the auth backends.",
+			},
+		},
+	}
+}
+
+func authBackendsDataSourceRead(d *schema.ResourceData, meta interface{}) error {
+	client, e := provider.GetClient(d, meta)
+	if e != nil {
+		return e
+	}
+
+	targetType := d.Get("type").(string)
+
+	auths, err := client.Sys().ListAuth()
+	if err != nil {
+		return fmt.Errorf("error reading from Vault: %s", err)
+	}
+
+	var paths, accessors []string
+
+	for path, auth := range auths {
+
+		path = strings.TrimSuffix(path, "/")
+
+		if targetType == "" {
+			paths = append(paths, path)
+			accessors = append(accessors, auth.Accessor)
+			// we do this to make test assertions easier
+			// this is not required
+			sort.Strings(paths)
```

Comment:
This doesn't seem to be required for the tests to pass. I don't know why I added it 🤔 . Can we remove it?

Reply:
I'll remove it and re-run the tests to be sure. The tests depended on an order, not sure if that will upset tests; will verify

Reply:
```text
--- FAIL: TestDataSourceAuthBackends (2.04s)
    data_source_auth_backends_test.go:22: Step 2/3 error: Check failed: Check 2/6 error: data.vault_auth_backends.test: Attribute 'paths.0' expected "approle", got "userpass"
FAIL
FAIL    github.com/hashicorp/terraform-provider-vault/vault     2.385s
FAIL
make: *** [testacc] Error 1
```

So I can update comparison order, however is it possible that the order would change based on the testing conditions I've supplied?

Reply:
So the first time I run the test, it fails, but subsequent runs pass successfully (with the sort removed)

Reply:
Re-added the sort to ensure tests are consistent. Unless you think it's better to sort only in the test, I'll leave this as-is for now. LMK if you think this is OK and resolved

Reply:
I think it is better to sort only in the test. But if we can't do that then I don't see a problem with leaving this as-is. I will leave it up to you.

Reply:
I think this is probably better since the sort operation isn't required in the source. Only thing I would suggest is to add a comment on why you are sorting in the test. Thanks!

Reply:
OK, getting this pushed in